paddle/fluid/framework/details/gather_op_handle_test.cc

```diff
@@ -0,0 +1,227 @@
+    //        send_vector, *(ctxs_[input_scope_idx]), in_lod_tensor);
+    //    in_lod_tensor->set_lod(lod);
+    //
+    //    gather_op_handle_->Run(false);
+    //
+    //    WaitAll();
+    //
+    //    p::CPUPlace cpu_place;
+    //    for (size_t j = 0; j < gpu_list_.size(); ++j) {
+    //      auto out_var = local_scope_[j]->Var("out");
+    //      auto out_tensor = out_var->Get<f::LoDTensor>();
+    //      PADDLE_ENFORCE_EQ(out_tensor.lod(), lod, "lod is not equal.");
+    //
+    //      f::Tensor result_tensor;
+    //      f::TensorCopy(out_tensor, cpu_place, *(ctxs_[j]), &result_tensor);
+    //      float* ct = result_tensor.mutable_data<float>(cpu_place);
+    //
+    //      for (int64_t j = 0; j < f::product(kDims); ++j) {
+    //        ASSERT_NEAR(ct[j], send_vector[j], 1e-5);
+    //      }
+    //    }
+    //
+    //    GatherOpDestroy();
+  }
+
+  void TestGatherSelectedRows() {
+    int output_scope_idx = 0;
+    InitGatherOp<f::SelectedRows>(output_scope_idx);
+
+    int height = kDims[0] * 2;
+    std::vector<int64_t> rows{0, 1, 2, 3, 3, 0, 14, 7, 3, 1,
+                              2, 4, 6, 3, 1, 1, 1,  1, 3, 7};
+    std::vector<float> send_vector(f::product(kDims));
+    for (size_t k = 0; k < send_vector.size(); ++k) {
+      send_vector[k] = k;
+    }
+
+    for (size_t input_scope_idx = 0; input_scope_idx < gpu_list_.size();
+         ++input_scope_idx) {
+      auto in_var = local_scope_[input_scope_idx]->Var("input");
+      auto in_selected_rows = in_var->GetMutable<f::SelectedRows>();
+      auto value = in_selected_rows->mutable_value();
+      value->mutable_data<float>(kDims, gpu_list_[input_scope_idx]);
+
+      in_selected_rows->set_height(height);
+      in_selected_rows->set_rows(rows);
+
+      paddle::framework::TensorFromVector<float>(
+          send_vector, *(ctxs_[input_scope_idx]), value);
+      value->Resize(kDims);
+    }
+
+    gather_op_handle_->Run(false);
+
+    WaitAll();
+
+    p::CPUPlace cpu_place;
+
+    auto out_var = local_scope_[output_scope_idx]->Var("out");
+    auto& out_select_rows = out_var->Get<f::SelectedRows>();
+    auto rt = out_select_rows.value();
+
+    PADDLE_ENFORCE_EQ(out_select_rows.height(), height, "height is not equal.");
+    for (size_t k = 0; k < out_select_rows.rows().size(); ++k) {
+      PADDLE_ENFORCE_EQ(out_select_rows.rows()[k], rows[k % rows.size()]);
+    }
+
+    f::Tensor result_tensor;
+    f::TensorCopy(rt, cpu_place, *(ctxs_[output_scope_idx]), &result_tensor);
+    float* ct = result_tensor.data<float>();
+
+    for (int64_t j = 0; j < f::product(kDims); ++j) {
+      ASSERT_NEAR(ct[j], send_vector[j % send_vector.size()], 1e-5);
+    }
+
+    GatherOpDestroy();
+  }
+
+ public:
+  f::Scope g_scope_;
+  std::vector<p::DeviceContext*> ctxs_;
+  std::vector<f::Scope*> local_scope_;
+  std::vector<p::Place> gpu_list_;
+  f::details::GatherOpHandle* gather_op_handle_;
+};
+
+// TEST_F(GatherTester, TestCPUGatherTestLodTensor) {
```

Comment:
clean up?

Reply:
`GatherLodTensor`  is also necessary when using multi-threads runs the `parallel_exe` on CPU but not GPU.
I am developing it.